fee-market/src/fee_operations.rs

```diff
@@ -20,27 +23,25 @@ pub trait FeeOperationsModule:
     fn distribute_fees(
         &self,
         hash_of_hashes: ManagedBuffer,
-        address_percentage_pairs: MultiValueEncoded<MultiValue2<ManagedAddress, usize>>,
+        operation: DistributeFeesOperation<Self::Api>,
     ) {
-        self.require_setup_complete();
-        let pairs = self.parse_pairs(address_percentage_pairs);
-        let opt_pairs_hash = self.generate_pairs_hash(&pairs, &hash_of_hashes);
-        if opt_pairs_hash.is_none() {
+        let operation_hash = operation.generate_hash();
+        if let Some(error_message) = self.validate_operation_hash(&operation_hash) {
+            self.complete_operation(&hash_of_hashes, &operation_hash, Some(error_message));
             return;
-        }
-        let pairs_hash = opt_pairs_hash.unwrap();
-        self.lock_operation_hash_wrapper(&hash_of_hashes, &pairs_hash);
+        };
+
+        self.require_setup_complete_with_event(&hash_of_hashes, &operation_hash);
```

Comment:
- please put this after `operation.generate_hash()`
- this check should not panic

Reply:
Same for other endpoints